### PR TITLE
cdc: back pressure scan speed

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

There will be a series of PRs aims to fix the TiKV CDC OOM issue (too many pending send messages are buffered in memory).

To resolve the issue, we have two strategies,
1. Limit the speed of producing messages during the initial scan stage. #9948 
2. Set an upper bound of pending send messages. #9885 

The series of PRs adopts both strategies.

This is part of #9885, it backpressure scan speed by replacing an unbounded channel with a bounded channel.

Plan:
- [x] take1: cherry-pick speed limiter. #10108 
- [ ] take2: set an upper bound of an internal channel used in CDC <-- we are here.
- [ ] take3: add memory quota to CDC and kill connection if it uses too much memory.
- [ ] take4: cherry-pick other bugfixes in #9885 

See full changes: https://github.com/tikv/tikv/compare/master...overvenus:cdc/flowcontrol/all

Cc #8168 #9996

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Support back pressure CDC scan speed.

### CODE REVIEW

**Comment by overvenus** [comment]
Cc @liuzix

File: components/cdc/src/channel.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+use tikv_util::impl_display_as_debug;
+
+use crate::service::{CdcEvent, EventBatcher};
+
+const CDC_MSG_MAX_BATCH_SIZE: usize = 128;
+// Assume the average size of event is 1KB.
+// 2 = (CDC_MSG_MAX_BATCH_SIZE * 1KB / CDC_EVENT_MAX_BATCH_SIZE).ceil() + 1 /* reserve for ResolvedTs */;
+pub const CDC_EVENT_MAX_BATCH_SIZE: usize = 2;
+
+pub fn canal(buffer: usize) -> (Sink, Drain) {
```
Comment (5kbpers) [inline components/cdc/src/channel.rs:22]: typo

File: components/cdc/src/channel.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            bounded_receiver,
+        },
+    )
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SendError {
+    Full,
+    Disconnected,
+    Congest,
```
Comment (5kbpers) [inline components/cdc/src/channel.rs:41]: Should be `Congested`?

File: components/cdc/src/channel.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+impl std::error::Error for SendError {}
+
+impl_display_as_debug!(SendError);
+
+macro_rules! impl_from_future_send_error {
+    ($($f:ty,)+) => {
+        $(
+            impl From<$f> for SendError {
+                fn from(e: $f) -> Self {
+                    if e.is_disconnected() {
```
Comment (5kbpers) [inline components/cdc/src/channel.rs:53]: How about using pattern matching like 
```rust
match e  {
    $f::Disconnected => {...},
}
```

File: components/cdc/src/delegate.rs [hunk 4]
```diff
@@ -993,113 +1000,4 @@ mod tests {
         assert!(err.has_epoch_not_match());
         assert!(err.take_epoch_not_match().current_regions.is_empty());
     }
-
-    #[test]
-    fn test_scan() {
```
Comment (5kbpers) [inline components/cdc/src/delegate.rs:998]: Should we add this test back later?

File: components/cdc/src/endpoint.rs [hunk 5]
```diff
@@ -1132,13 +1149,16 @@ impl Initializer {
...
             debug!("cdc scan entries"; "len" => entries.len(), "region_id" => region_id);
             fail_point!("before_schedule_incremental_scan");
-            let scanned = Task::IncrementalScan {
-                region_id,
-                downstream_id,
-                entries,
-            };
-            if let Err(e) = self.sched.schedule(scanned) {
-                error!("cdc schedule cdc task failed"; "error" => ?e, "region_id" => region_id);
+            if let Err(e) = self.sink_scan_events(entries, done, require_barrier).await {
```
Comment (5kbpers) [inline components/cdc/src/endpoint.rs:1152]: If we send the scan event here, then ticdc could still receive scanned entries after a region had been deregistered. Is that OK?

**Review by 5kbpers — COMMENTED** [review]
rest LGTM

File: components/cdc/src/channel.rs (hunk 3 above)
Comment (overvenus) [reply components/cdc/src/channel.rs:53]: `futures::channel::mpsc::{SendError, TrySendError}` are not enums. 😅

File: components/cdc/src/delegate.rs (hunk 4 above)
Comment (overvenus) [reply components/cdc/src/delegate.rs:998]: It tests `on_scan` but it is removed in this PR, so I think we can just remove the test too.

File: components/cdc/src/endpoint.rs (hunk 5 above)
Comment (overvenus) [reply components/cdc/src/endpoint.rs:1152]: @amyangfei PTAL

File: components/cdc/src/channel.rs (hunk 3 above)
Comment (overvenus) [reply components/cdc/src/channel.rs:53]: See more: https://docs.rs/futures/0.3.14/futures/channel/mpsc/struct.SendError.html

**Pull request merged**